src/os/filestore/FileStore.cc

```diff
@@ -237,18 +237,22 @@ int FileStore::lfn_open(const coll_t& cid,
 			const ghobject_t& oid,
 			bool create,
 			FDRef *outfd,
-                        Index *index)
+			Index *index,
+			bool directio)
 {
   assert(outfd);
   int r = 0;
   bool need_lock = true;
   int flags = O_RDWR;
+  FDCache *cachep = directio ? &fdcache_direct : &fdcache;
 
   if (create)
     flags |= O_CREAT;
   if (g_conf->filestore_odsync_write) {
     flags |= O_DSYNC;
   }
+  if (directio)
+    flags |= O_DIRECT;
```

Comment:
Hmm, for some reason, I assumed it was ok to open with both.  This is actually a problem.  I think we need to not cache DIRECT opened fds.  That means that lfn_open should bail with an error if the fd is currently cached.  If it's not cached, it needs to return an FDRef which is not cached, while making sure to block any other lfn_open callers on the same hobject until the ref is released (in order to ensure that no one else tries to open that file while the read is happening).  This is going to add a fair degree of complexity, I'm afraid.